File: rpsls_j17/src/test/java/nl/bentels/rpsls/GameTest.java
package nl.bentels.rpsls;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class GameTest {

    private static final ByteArrayOutputStream BAOS = new ByteArrayOutputStream();

    @BeforeAll
    public static void swapIOStreams() {
        System.setOut(new PrintStream(BAOS));
        System.setIn(new ByteArrayInputStream("spock".getBytes(StandardCharsets.UTF_8)));
    }

    private Game objectUnderTest;

    @BeforeEach
    public void createTestObject() {
        objectUnderTest = new Game(new Spock());
    }

    @Test
    public void whenGameRun_thenInteractionOkay() {
        objectUnderTest.doGame();

        String[] strings = BAOS.toString().split("\n");
        assertEquals(3, strings.length);
        assertEquals("Which role do you want (ROCK, PAPER, SCISSORS, LIZARD, SPOCK)? You chose Spock", strings[0]);
        assertEquals("I chose Spock", strings[1]);
        assertEquals("It's a draw", strings[2]);

    }

}

File: rpsls_grdl/app/src/test/kotlin/nl/bentels/rpsls/GameTest.kt
package nl.bentels.rpsls;

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.io.*

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class GameTest {

    val bos = ByteArrayOutputStream()

    @BeforeEach
    fun initIO (){
        println("Initializing...")
        val spock = PlayerRole.SPOCK.name.byteInputStream()
        System.setIn(spock)
        System.setOut(PrintStream(bos))
    }

    @Test
    fun whenGameExercisedThenOkay() {
        Game(PlayerRole.SPOCK).playGame()
        val resStrings = bos.toString().split('\n')
        print(resStrings)
        assertEquals(4, resStrings.size)
        assertEquals("Which role do you want (SCISSORS, PAPER, ROCK, LIZARD, SPOCK)? You chose SPOCK", resStrings[0])
        assertEquals("I chose SPOCK", resStrings[1])
        assertEquals("It's a DRAW", resStrings[2])
    }

}
